### PR TITLE
feat: add stripe authentication support

### PR DESCRIPTION
**Pull request by ahmaxed**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #XXXXX

### CODE REVIEW

**Review by naomi-lgbt — CHANGES_REQUESTED** [review]
I'm nitpicking 🙂

**Comment by ahmadabdolsaheb** [comment]
@nhcarrigan, thanks for that.

File: client/src/components/Donation/stripe-card-form.tsx [hunk 1]
```diff
@@ -9,19 +9,45 @@ import {
...
+        }
+    >
+  | { error: { type: 'StripeNotLoaded' } };
+
 interface FormPropTypes {
   onDonationStateChange: (donationState: AddDonationData) => void;
-  postStripeCardDonation: (token: Token) => void;
+  postStripeCardDonation: (
+    paymentMethodId: string,
+    handleAuthentication: handleAuthenticationType
```
Comment (ojeytonwilliams) [inline client/src/components/Donation/stripe-card-form.tsx:49]:
```suggestion
    handleAuthentication: HandleAuthentication
```
I'll get around to adding a lint rule one of these days...

File: client/src/components/Donation/DonateForm.tsx [hunk 2]
```diff
@@ -29,7 +29,7 @@ import Spacer from '../helpers/spacer';
 import DonateCompletion from './DonateCompletion';
 import type { AddDonationData } from './PaypalButton';
 import PaypalButton from './PaypalButton';
-import StripeCardForm from './stripe-card-form';
+import StripeCardForm, { handleAuthenticationType } from './stripe-card-form';
```
Comment (ojeytonwilliams) [inline client/src/components/Donation/DonateForm.tsx:32]:
```suggestion
import StripeCardForm, { HandleAuthentication } from './stripe-card-form';
```

File: client/src/components/Donation/DonateForm.tsx [hunk 3]
```diff
@@ -57,9 +57,10 @@ type DonateFormProps = {
   addDonation: (data: unknown) => unknown;
   postChargeStripe: (data: unknown) => unknown;
   postChargeStripeCard: (data: {
-    token: Token;
+    paymentMethodId: string;
     amount: number;
     duration: string;
+    handleAuthentication: handleAuthenticationType;
```
Comment (ojeytonwilliams) [inline client/src/components/Donation/DonateForm.tsx:63]:
```suggestion
    handleAuthentication: HandleAuthentication;
```

File: client/src/components/Donation/DonateForm.tsx [hunk 4]
```diff
@@ -221,17 +222,21 @@ class DonateForm extends Component<DonateFormProps, DonateFormComponentState> {
     });
   }
 
-  postStripeCardDonation(token: Token) {
+  postStripeCardDonation(
+    paymentMethodId: string,
+    handleAuthentication: handleAuthenticationType
```
Comment (ojeytonwilliams) [inline client/src/components/Donation/DonateForm.tsx:227]:
```suggestion
    handleAuthentication: HandleAuthentication
```

File: client/src/redux/donation-saga.js [hunk 5]
```diff
@@ -73,10 +74,45 @@ function* postChargeStripeSaga({ payload }) {
...
+  clientSecret,
+  paymentMethodId
+) {
+  if (error.type === 'UserActionRequired' && clientSecret) {
+    yield handleAuthentication(clientSecret, paymentMethodId)
+      .then(result => {
+        if (
+          !result.paymentIntent?.status ||
+          !result.paymentIntent.status === 'succeeded'
+        )
```
Comment (ojeytonwilliams) [inline client/src/redux/donation-saga.js:89]:
```suggestion
        if (result?.paymentIntent?.status !== 'succeeded')
```
It's supposed to throw unless the status is 'succeeded', no?

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]

File: client/src/components/Donation/stripe-card-form.tsx (hunk 1 above)
Comment (ShaunSHamilton) [reply client/src/components/Donation/stripe-card-form.tsx:49]: The Lint rule will become annoying with externally typed declarations. E.g. the editor (monaco)

Comment (ojeytonwilliams) [reply client/src/components/Donation/stripe-card-form.tsx:49]: It shouldn't apply to them - we don't lint packages, only our code.

Comment (ShaunSHamilton) [reply client/src/components/Donation/stripe-card-form.tsx:49]: But, we import the types.

Comment (ShaunSHamilton) [reply client/src/components/Donation/stripe-card-form.tsx:49]: For the editor, I cannot think of anything `SoAndSoType`, but you also have an aversion to `ISoAndSo`

Comment (ojeytonwilliams) [reply client/src/components/Donation/stripe-card-form.tsx:49]: > But, we import the types.

Doesn't matter.  The linter would only check our definitions, not the imported ones.

Comment (ShaunSHamilton) [reply client/src/components/Donation/stripe-card-form.tsx:49]: Ah, well then ignore me.

File: api-server/src/server/utils/donation.js [hunk 6]
```diff
@@ -221,26 +218,51 @@ export async function createStripeCardDonation(req, res, stripe) {
...
+    subscriptionId = subscription_id;
+  } catch (err) {
+    if (
+      err.type === 'UserActionRequired' ||
+      err.type === 'PaymentMethodRequired'
+    )
+      throw err;
+    else
+      throw {
+        type: 'subscriptionCreationFailed',
```
[the remark this answers is not in the document]
Comment (ahmadabdolsaheb) [reply api-server/src/server/utils/donation.js:263]: Yes we will handle them in an upcoming pr that uses the client's i18n support to display the errors.

File: api-server/src/server/utils/donation.js [hunk 7]
```diff
@@ -209,8 +205,9 @@ export async function createStripeCardDonation(req, res, stripe) {
   let customerId;
   try {
     const customer = await stripe.customers.create({
-      email,
-      card: tokenId,
+      email: 'test2@test.org',
```
[the remark this answers is not in the document]
Comment (ahmadabdolsaheb) [reply api-server/src/server/utils/donation.js:208]: yes, will be removed. Thanks for catching.

**Comment by ojeytonwilliams** [comment]
Thanks @ahmadabdolsaheb, the code looks good.  Try and get the QAing done tomorrow.

**Comment by ahmadabdolsaheb** [comment]
Here are the test cards: https://stripe.com/docs/billing/subscriptions/fixed-price#test

**Review by scissorsneedfoodtoo — APPROVED** [review]
Took a look at the code, but there were already a couple of great reviews, and I didn't notice anything outside of what was mentioned in those.

Tested the 3D secure authentication with the second and third cards in the link that was shared above. Both cards triggered the 3D secure popup and succeeded or failed due to insufficient funds as expected.

Also checked things out in the Stripe test mode console and the subscriptions from my testing all show up as expected.

Everything LGTM :+1: :+1:

**Comment by ojeytonwilliams** [comment]
In that case, I'm happy to merge this and test on staging.  @nhcarrigan do you want to take a final look?

**Review by ojeytonwilliams — APPROVED** [review]
LGTM :+1:

**Review by naomi-lgbt — APPROVED** [review]
Looks good to me

**Pull request merged**